src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityCreateController.php

```diff
@@ -79,23 +81,94 @@ public function typeAction(Request $request, $targetEnvironment)
     }
 
     /**
-     * The create action serves two routes.
-     *
-     *  1. entity_add to create new entities.
+     * @Method({"GET", "POST"})
+     * @Route("/entity/create/{type}/{targetEnvironment}", name="entity_add")
+     * @Template("@Dashboard/EntityEdit/edit.html.twig")
+     * @param Request $request
+     * @param null|string $targetEnvironment
+     * @param null|string $type
      *
-     *  2. entity_copy to copy entities from Manage into the SP Dashboard.
-     *     The Entity is copied from the manage test environment to become either a SP Dashboard test-draft or
-     *     production-draft entity.
+     * @return RedirectResponse|Response
      *
+     * @SuppressWarnings(PHPMD.NPathComplexity)
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     * @SuppressWarnings(PHPMD.ElseExpression)
+     */
+    public function createAction(Request $request, $targetEnvironment, $type)
+    {
+        $flashBag = $this->get('session')->getFlashBag();
+        $flashBag->clear();
+
+        $service = $this->getService();
+
+        if (!$service->isProductionEntitiesEnabled() &&
+            $targetEnvironment !== Entity::ENVIRONMENT_TEST
+        ) {
+            throw $this->createAccessDeniedException(
+                'You do not have access to create entities without publishing to the test environment first'
+            );
+        }
+
+        if ($type == Entity::TYPE_OPENID_CONNECT) {
```

Comment:
This might be a good moment to move creation of the commands to a factory

Reply:
A good suggestion